src/solvers/dg/2d/dg.jl

```diff
@@ -1879,11 +1883,135 @@ end
+                           nonconservative_terms::Val{true}, mortars, cache)
+  @unpack neighbor_ids, u_lower, u_upper, orientations = mortars
+  @unpack fstar_upper_threaded, fstar_lower_threaded = cache
+
+  # Type alias only for convenience
+  A2d = MArray{Tuple{nvariables(dg), nnodes(dg)}, Float64}
+
+  noncons_diamond_upper_threaded = [A2d(undef) for _ in 1:Threads.nthreads()]
+  noncons_diamond_lower_threaded = [A2d(undef) for _ in 1:Threads.nthreads()]
+
+  Threads.@threads for m in 1:dg.n_l2mortars
+    large_element_id = dg.l2mortars.neighbor_ids[3, m]
+    upper_element_id = dg.l2mortars.neighbor_ids[2, m]
+    lower_element_id = dg.l2mortars.neighbor_ids[1, m]
+
+    # Choose thread-specific pre-allocated container
+    fstar_upper = fstar_upper_threaded[Threads.threadid()]
+    fstar_lower = fstar_lower_threaded[Threads.threadid()]
+
+    noncons_diamond_upper = noncons_diamond_upper_threaded[Threads.threadid()]
+    noncons_diamond_lower = noncons_diamond_lower_threaded[Threads.threadid()]
+
+    # Calculate fluxes
+    calc_fstar!(fstar_upper, u_upper, m, orientations, dg)
+    calc_fstar!(fstar_lower, u_lower, m, orientations, dg)
+
+    # Compute the nonconservative numerical terms along the upper and lower interface
+    # Done twice because left/right orientation matters
+    # 1 -> primary element and 2 -> secondary element
+    # See Bohm et al. 2018 for details on the nonconservative diamond "flux"
+    if dg.l2mortars.large_sides[m] == 1 # -> small elements on right side
+      for i in 1:nnodes(dg)
+        # pull the left and right solutions
+        u_upper_ll, u_upper_rr = get_surface_node_vars(u_upper, dg, i, m)
+        u_lower_ll, u_lower_rr = get_surface_node_vars(u_lower, dg, i, m)
+        # Call pointwise nonconservative term
+        noncons_upper = noncons_interface_flux!(u_upper_ll,u_upper_rr,orientations[m],equations(dg))
+        noncons_lower = noncons_interface_flux!(u_lower_ll,u_lower_rr,orientations[m],equations(dg))
+        # Save to primary and secondary temporay storage
+        for v in 1:nvariables(dg)
+          noncons_diamond_upper[v, i] = noncons_upper[v]
+          noncons_diamond_lower[v, i] = noncons_lower[v]
+        end
+      end
+    else # large_sides[m] == 2 -> small elements on the left
+      for i in 1:nnodes(dg)
+        # pull the left and right solutions
+        u_upper_ll, u_upper_rr = get_surface_node_vars(u_upper, dg, i, m)
+        u_lower_ll, u_lower_rr = get_surface_node_vars(u_lower, dg, i, m)
+        # Call pointwise nonconservative term
+        noncons_upper = noncons_interface_flux!(u_upper_rr,u_upper_ll,orientations[m],equations(dg))
+        noncons_lower = noncons_interface_flux!(u_lower_rr,u_lower_ll,orientations[m],equations(dg))
+        # Save to primary and secondary temporay storage
+        for v in 1:nvariables(dg)
+          noncons_diamond_upper[v, i] = noncons_upper[v]
+          noncons_diamond_lower[v, i] = noncons_lower[v]
+        end
```

Comment:
```suggestion
        set_node_vars!(noncons_diamond_upper, noncons_upper, dg, i)
        set_node_vars!(noncons_diamond_lower, noncons_lower, dg, i)
```
To consistently use `get/set_node_vars`.

Reply:
fixed. Also should I use `set_node_vars!`in lines 1820 - 1828 in the `calc_interface_flux!`routine?

Reply:
ah nevermind the sizes do not match